src/core/inductive.ml

```diff
@@ -0,0 +1,257 @@
+  | Appl(t1, t2) -> Pos.none (P_Appl (type_to_pattern t1, type_to_pattern t2))
+  | Prod _       -> fatal None "Prod - Not yet implemented"
+  | Abst _       -> fatal None "Abst - Not yet implemented"
+  | LLet _       -> fatal None "LLet - Not yet implemented"
+  | Type         -> fatal None "TYPE - Not yet implemented"(*Pos.none P_Type*)
+  | Kind         -> fatal None "[Kind] not possible"
+  | Wild         -> fatal None "[Wild] not possible"
+  | TRef _       -> fatal None "[TRef] not possible"
+  | Meta _       -> fatal None "[Meta] not possible"
+  | Patt _       -> fatal None "[Patt] not possible"
+  | TEnv _       -> fatal None "[TEnv] not possible"
+
+(** Some constructors to create a p_term without position *)
+
+let _P_Iden : qident -> bool -> p_term = fun ident b ->
+  Pos.none (P_Iden(ident, b))
+
+let _P_Patt : strloc option -> p_term array -> p_term = fun s ts ->
+  Pos.none (P_Patt(s, Some ts))
+
+let _P_Appl : p_term -> p_term -> p_term = fun t1 t2 ->
+  Pos.none (P_Appl(t1, t2))
+
+(** [create_patt t] creates a pattern $[t] without position *)
+let create_patt : strloc -> p_term = fun t -> _P_Patt (Some t) [||]
+
+(*
+let create_arg : int -> p_term list = fun len ->
+  let rec aux : int -> p_term list -> p_term list = fun i acc ->
+    if i = 0 then
+      acc
+    else
+      let head =
+        Pos.none (P_Patt (Some (Pos.none ("a" ^ (string_of_int i))), None))
+      in
+      aux (i-1) (head::acc)
+  in
+  aux len []*)
+
+(** [rec_rules ind_sym rec_sym cons_list] returns the p_rules linking with an
+    induction principle [rec_sym] of the inductive type [ind_sym] (with its
+    constructors [cons_list]). *)
+let rec_rules : sym -> sym -> sym list -> p_rule list =
+  fun ind_sym rec_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let app   : p_term -> p_term list -> p_term = List.fold_left _P_Appl in
+  let fapp  : p_term list -> p_term -> p_term = fun ts f -> app f ts in
+(*  let app2  : sym -> sym list -> p_term = fun t q ->
+    let p_patt = create_patt (Pos.none ("pi" ^ t.sym_name)) in
+    List.fold_left (fun q -> _P_Appl q p_patt)
+  in
+  let fapp2 : p_term -> sym list -> p_term = fun p ts -> app2 p ts  in*)
+
+
+  (* STEP 1: Create the common head of the rules *)
+
+  (*let p_iden = _P_Iden (Pos.none ([], sym_ind.sym_name)) true in
+  let p_patt = create_patt (Pos.none "p")                     in
+  let head =
+    List.map (fun c -> create_patt (Pos.none ("pi" ^ c.sym_name))) cons_list
+  in
+  let common_head = fapp head (_P_Appl p_iden p_patt) in*)
+  let p_iden = _P_Iden (Pos.none ([], rec_sym.sym_name)) true in
+  let p_patt = create_patt (Pos.none "p") in
+  let head   = _P_Appl p_iden p_patt in
+  let arg : sym list -> p_term = fun l  ->
+    let rec aux : sym list -> p_term -> p_term = fun l acc ->
+      match l with
+      | []   -> acc
+      | t::q ->
+          let p_patt = create_patt (Pos.none ("p" ^ t.sym_name)) in
```

Comment:
This may create problems if constructor names are not regular identifiers. To avoid this kind of problems now, it is better to use names like p0, p1, etc.

Reply:
There is no problem with regular identifiers (See my recent test), and it's more readable than p0 p1 etc.
Moreover, there is the function `add_prefix`.

Reply:
Your code should not depend on the fact that identifiers are regular or not. If one constructor is a declared ident, these variables should be declared idents too, which requires to update the signature state dynamically. This is too complicated. So, please use p1 .. pn and x1 .. xk for the constructor arguments. Thank you.